librad/src/id/data.rs

```diff
@@ -1,13 +1,21 @@
-use serde::{Deserialize, Serialize};
-use std::collections::{HashMap, HashSet};
-//use serde_json;
-
+use crate::id::Error;
+use olpc_cjson::CanonicalFormatter;
+use serde::{Deserialize, Serialize, Serializer};
+use std::collections::{BTreeSet, HashMap, HashSet};
 #[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
 pub struct EntitySignatureData {
     pub user: Option<String>,
     pub sig: String,
 }
 
+fn ordered_set<S>(value: &HashSet<String>, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    let ordered: BTreeSet<String> = value.iter().cloned().collect();
+    ordered.serialize(serializer)
+}
+
 #[derive(Clone, Deserialize, Serialize, Debug, Default)]
 pub struct EntityData {
```

Comment:
I'm actually a bit confused as to why most of the fields here are `Option`, could you give me a run down?